File: object-c/08_stack/stack_practice/ArrayStack.m
#import "ArrayStack.h"

@implementation Stack {
    NSMutableArray *_arr;
    NSUInteger _capacity;
    NSUInteger _count;
}

- (id)initWithCapacity:(NSUInteger)capacity {
    self = [super init];
    _capacity = capacity;
    _arr = [[NSMutableArray alloc] initWithCapacity:capacity];
    return self;
}

- (BOOL)isEmpty {
    return _arr.count == 0;
}

- (BOOL)isFull {
    return _arr.count == _capacity;
}

- (id)top {
    if ([self isEmpty]) return nil;
    NSUInteger index = _arr.count - 1;
    return _arr[index];
}

- (NSUInteger)size {
    return _arr.count;
}

- (BOOL)push:(id)obj {
    if (!obj) return NO;
    if (_arr.count == _capacity) return NO;
    [_arr addObject:obj];
    return YES;
}

- (id)pop {
    if ([self isEmpty]) return nil;
    NSUInteger index = _arr.count - 1;
    id obj = _arr[index];
    [_arr removeLastObject];
    return obj;
}

@end

File: csharp/08-stack/algo08_stack/ArrayStack.cs
using System;

namespace algo08_stack
{
    public class ArrayStack<T>
    {
        private readonly int _capacity;

        private readonly T[] _data;

        private int _top = -1; // 指向栈顶元素，当为-1时表示栈为空

        public ArrayStack(int capacity)
        {
            _capacity = capacity;

            _data = new T[capacity];
        }

        public int Count => _top + 1;

        public void Push(T val)
        {
            if (Count == _capacity) throw new InvalidOperationException("Stack full.");

            _top++;

            _data[_top] = val;
        }

        public T Pop()
        {
            if (_top == -1) throw new InvalidOperationException("Stack empty.");

            T val = _data[_top];
            _top--;

            return val;
        }
    }
}